### ISSUE TITLE
Survey.com

### ISSUE THREAD

**Issue by b1ackmai1er** [wwarthen/RomWBW#316]
Hi Wayne,

Survey.com crashes when run on an ecb-sbc-v2 (port 77 I think) and this leads people to beleive there is an issue with their device.
Suggest at a minimum it should display a warning message that this could occur.

Ideally survey should be updated to recognize the platform and have a i/o map of ports it should skip.

Maybe at the same time it could be converted to Zilog mnemonics and made to build under TASM.

Another option maybe to have a utility that iterates through all the devices HBIOS has identified and displays them.

Does survey.com actually have any real value as being part of the base install. Is there a alternate program prgram perhaps?

Thanks

**Comment by wwarthen** [comment]
I finally decided to hack SURVEY.COM for now.  Seems like folks do use it.  The hacked version is checked in to the dev branch.

Over time, it should be replaced with something better, but this hack should allow it to avoid the bank switching issues.  The code now requires RomWBW and I updated the banner to reflect that this is a RomWBW-specific variant.  I didn't add a RomWBW check in the code, so it will just crash on non-RomWBW systems.  Didn't see  like it was worth the effort.

Note that it still doesn't detect the SBC bank switching ports as active.  This is because the ports do not assert anything on the data bus during a read operation.

I will wait a few days before closing this issue to see if there are any problems with my hack.

-Wayne

**Comment by wwarthen** [comment]
Guess I'll go ahead and close this.  Will reopen if there are related issues.

-Wayne

**Comment by b1ackmai1er** [comment]
Thanks Wayne, confirming no crash on my ecb-sbc-v2

How does the detection even work? Is it only valid for i/o ports on the cpu
side of the data bus drivers?
Why do no high ports get detected?

Thanks

Oh yeas, good hack, i didn't even know that you could switch 8080/z80
mnemonics inline.

[image: image.png]
Why Why Why

**Comment by wwarthen** [comment]
> How does the detection even work? Is it only valid for i/o ports on the cpu side of the data bus drivers? Why do no high ports get detected?

Guess I closed this too soon.  The open port detection is based on some loose heuristics.  The algorithm just assumed that the values 0x78 and 0xFF mean that the port is unused.  It also assumed that if the value returned is equal to the port number, then the port is inactive.  Those values *can* all indicate an open port depending on the bus configuration.

Unfortunately, it was failing badly on the SBC for some reason.  I was seeing odd port values like 0xFE from that bus which was being interpreted as an active port when it was not.

I spend a few hours on this today and will be checking in a new iteration momentarily.  Basically, I changed the algorithm to probe (read) the port using two different instructions (IN A,<port> and IN A,(C)).  I have seen this trick before.  For an open port, the values returned will almost always be different for the different instructions.  Additionally, the value 0xFF is considered an open port even if both instructions return the same value.  This handles the case where the data lines have a weak pull-up applied.

In addition to the updated SURVEY.COM, I am also checking in a new test app called PORTSWP.COM.  It stands for port sweep.  This app just scans all ports and, for each port, dumps the value returned using both  instructions.  It then does both again resulting in 4 values total for each port.  This gives a good idea of what any given port is really doing and how consistent the results are.

Both tools go through the brain damage of running in high memory and restoring a possibly corrupt bank selection after each probe.  This allows it to run properly and successfully on the SBC.  The tool also need to differentiate between Z80 and Z180 CPUs because on a Z180 it is mandatory to use IN0 <port> instead of IN <port>.  Ridiculous amount of work for such a simple thing.

By the way, the detection is still far from perfect because:

1. It will not detect "write only" ports (not possible as far as I know).
2. If an active port returns 0xFF as a true value, it will be considered inactive.
3. If a port is returning a running counter value if will be considered inactive because the two reads will return different values as the counter runs.

I don't see any way to work around these deficiencies.

I would be happy to get some feedback on how this is working.  Seems pretty good on my SBC and Mark IV.

-Wayne

**Comment by b1ackmai1er** [comment]
Hi Wayne,

New build of survey locks up for me after port 77 most of the time. It
either freezes, prints a garbage character or exit to cpm or prints
repetitive text.

Can try tracing it tomorrow if you cannot duplicate.

Regards Phil

**Comment by b1ackmai1er** [comment]
Note, that the new version does not get to the "ports active: message and
looking back at my previous screenshot of your last version, neither did
that.

Regards Phil

**Comment by wwarthen** [comment]
Thanks for testing Phil.  I have no idea what is happening here.  I just took the built version of `SURVEY.COM` from the last GitHub check-in and ran it on both my SBC and my Mark IV.  Both worked fine.  Below is what the output looks like from my SBC.  By the way, I am not seeing the screen shot that you attached previously.  Not showing  up in GitHub for some reason.  I want to clarify.  You **are** seeing the `Active I/O ports:` message prior to scanning, right?  You are **not** seeing the `26 Ports active` message at the conclusion of scanning, right?  I know it is a stupid question, but is there any chance you are running an old version?  Confirm that you see "RomWBW" in the banner of the program.

Thanks,

Wayne

```
        *** RomWBW System Survey (Feb 2023) ***

Drive A: 26K bytes in 1 files with 230K bytes remaining
Drive B: 348K bytes in 49 files with 36K bytes remaining

Memory map:
0       8       16      24      32      40      48      56      64
|       |       |       |       |       |       |       |       |
 TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRR
T=TPA   C=CPM   B=BIOS or unassigned    R=ROM or bad
BIOS at E603    iobyte 94       drive 00        BDOS at D806

33251 Bytes RAM         32284 Bytes ROM         55302 Bytes in TPA
0 Bytes Empty           65535 Total Active Bytes

Active I/O ports:
20 22 23 24 26 27
30 31 32 33 34 35 36 37
60 62 64 66 68 69 6A 6B 6C 6D 6F
78
26 Ports active
```

**Comment by wwarthen** [comment]
Oh, one other thing.  If it is still misbehaving for you, could you try the new `PORTSWP.COM` application and post the results.

Thanks,

Wayne

**Comment by wwarthen** [comment]
So, I think I just recreated the problem.  I tested my SBC-V004 and both `PORTSWP` and `SURVEY` crashed.  After a bit of sleuthing, I found that the VGM card seemed to be the source of the crash.  Removing the card allowed both apps to run to completion normally.  Additionally, `PORTSWP` showed that the crash occurs when it reached the SN76489 ports (0xC6-0xC7).

@b1ackmai1er, did your system have a VGM card installed?

I have no idea why the VGM card would cause this.

-Wayne

**Comment by wrljet** [comment]
Wayne,

I guess I would grow used to it, but PORTSWP to me sounds like Swap rather than Sweep.  How about PORTSCAN?

Bill

**Comment by wwarthen** [comment]
Strongly agree @wrljet.  Will make that change.  Thanks!

**Comment by wwarthen** [comment]
> I have no idea why the VGM card would cause this.

Just tried using the Monitor to (I)nput from the SN76489 ports.  Doing this on either of the ports is locking up my system.  Interesting and odd.

-Wayne

**Comment by wwarthen** [comment]
> Strongly agree @wrljet. Will make that change. Thanks!

dc151ad

**Comment by wwarthen** [comment]
Did a little more research on the VGM board and the crash that occurs during a port scan.  I think I see what is happening.

The SN76489 asserts /WAIT when it is not ready for the next command.  It starts asserting /WAIT as soon as it sees /OE asserted.  It then waits for a command (/WE) and deasserts /WAIT when it is done with the command.  In the case of the VGM, /OE will be asserted for either a read or write I/O.  In the case of a read, the command never comes along (never gets the /WE) and /WAIT remains asserted forever.

I looked at a similar circuit on the RC2014 SN76489 module.  It qualifies /OE with /WR from the CPU, so the SN76489 will never see a write I/O at all.  I suspect that something similar will be required for the VGM.

Let me know what you think @b1ackmai1er, but I am inclined to close this issue.  I don't think there is an issue with `SURVEY` at this point.  This just demonstrates that doing a port scan is a dangerous operation.

Thanks,

Wayne

**Comment by b1ackmai1er** [comment]
Hi Wayne, I think your conclusion is right.

Was able to replicate. Survey runs fine with the vgm card removed or with the onboard /WAIT jumper removed.

You were right, I was running the old version of survey. For some reason I couldn't get the new version included in my rom. I had to delete all copies of survey. Maybe a hiccup with Survey.com vs. survey.com. Anyway got there in the end.

I see how you it would have been easy enough to narrow the problem down to the vgm card but how did you figure out it was am issue with the /WR driving /WE? To me it looks correct to the data sheet. I would never have figured that out. Maybe it is correct but reading from the chip was never intended, tested or identified as an issue by TI. New version of boards are on the way, so will patch them with /CE driving /WR like the RC2014 boards and provide let you know if that works.

Great sleuthing, thanks.

**Comment by wwarthen** [comment]
> I see how you it would have been easy enough to narrow the problem down to the vgm card but how did you figure out it was an issue with the /WR driving /WE? To me it looks correct to the data sheet. I would never have figured that out. Maybe it is correct but reading from the chip was never intended, tested or identified as an issue by TI. New version of boards are on the way, so will patch them with /CE driving /WR like the RC2014 boards and let you know if that works.

Yes, at first when I looked at the schematic I also thought that looks fine.  After I thought about it a while, I realized that it had to have something to do with writes vs. reads and it was clearly an issue with /WAIT being left asserted.  I reviewed the data sheet.  It did not address the way that /WAIT worked exactly, but I saw some reference to /OE controlling something and realized that /OE might be all that is needed to assert /WAIT.  At that point, I tried to recreate the problem on RC2014 and could not.  Finally looked at the RC2014 schematic and saw the difference.  I think you are correct that it was never intended for the chip to see a read.

Anyway, I am inclined to say that `SURVEY` is finally working OK.  Unless I hear otherwise from you, I will go ahead and close this issue.

Thanks,

Wayne
